Use components for RobotPropertyKind

xiyuoh opened the pull request:
This PR is a follow up to https://github.com/open-rmf/rmf_site/pull/263, specifically this review [comment](https://github.com/open-rmf/rmf_site/pull/263#discussion_r1926727422) about avoiding excessive serialization of Robot property data. The previous approach opted to store all Robot properties and property kinds in `json` to avoid having to query multiple different components on load/save. This PR expanded on the plugin approach by triggering updates to components via the property plugin's systems and only serialize data when a value has changed.

Another change is the use of a reusable plugin template `InspectRobotPropertyKindPlugin` with widget systems for all Robot property kinds rather than having the widgets stored in a resource. This allows each of the property kind plugins to have better access to world variables, instead of having to pass variables through `ShowRobotPropertyWidgetFn` in the previous implementation.

mxgrey reviewed (changes requested):
This is a very good implementation of what I was suggesting.

I've left some inline comments. A few of them should be addressed before merging, but some are just potential opportunities to reduce boilerplate which could be considered for a future PR.

xiyuoh commented on rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_robot_properties.rs:
```diff
@@ -179,47 +210,254 @@ where
+        + RobotPropertyKind
+        + Component
+        + for<'de> Deserialize<'de>,
+    U: 'static + Send + Sync + Default + Clone + RobotProperty + Component,
```
This is a great tip, thanks! [00634b4](https://github.com/open-rmf/rmf_site/pull/265/commits/00634b422a8081749f4c61c0f57ab5f5f3e62f40)

xiyuoh commented on rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_robot_properties.rs:
```diff
@@ -179,47 +210,254 @@ where
-        let inspector = app.world.resource::<RobotPropertiesInspector>().id;
+/// Implement this plugin to add a new configurable robot property kind of type T to the
+/// robot properties inspector.
+pub struct InspectRobotPropertyKindPlugin<W, T, U>
```
Updated in [00634b4](https://github.com/open-rmf/rmf_site/pull/265/commits/00634b422a8081749f4c61c0f57ab5f5f3e62f40). I left `W` as it is because `Widget` is a key word and cannot be used, and `KindWidget` seems a little too long as a replacement 🤣 Since `W` has been consistently used in other parts of the code I assume it'd be sufficiently indicative.

xiyuoh commented on rmf_site_editor/src/widgets/inspector/inspect_model_description/mod.rs:
```diff
@@ -454,7 +448,7 @@ fn get_selected_description_entity<'w, 's, S: ReadOnlyWorldQuery, T: WorldQuery>
             return Some(affiliation);
         } else {
-            warn!("Model instance is affiliated with a non-existent description");
+            debug!("Model instance is affiliated with a description with different components");
```
Yes! Because this `get_selected_description_entity()` method is proving useful and called more frequently now. Removed in [00634b4](https://github.com/open-rmf/rmf_site/pull/265/commits/00634b422a8081749f4c61c0f57ab5f5f3e62f40)

xiyuoh commented on rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_mobility.rs:
```diff
@@ -39,10 +43,84 @@ pub struct InspectMobility<'w, 's> {
 }
 
 impl<'w, 's> WidgetSystem<Inspect> for InspectMobility<'w, 's> {
+    fn show(
```
Thanks for the insight! My goal was to put as much common code into `show_robot_property_widget()` for users to reuse, at the same time leaving space for them to add any custom logic within `show()`. Will look into this again.

xiyuoh commented on rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_collision.rs:
```diff
@@ -57,28 +137,83 @@ impl<'w, 's> WidgetSystem<Inspect> for InspectCollision<'w, 's> {
+        {
+            serialize_and_change_robot_property::<Collision, CircleCollision>(
+                params.change_robot_property,
+                CircleCollision::default(),
```
I apologise for the confusing implementation. This is called when a user has freshly enabled the `Collision` RobotProperty and selected the `CircleCollision` RobotPropertyKind. In this moment, both `Collision` and `CircleCollision` components would be inserted to the entity, but the `Collision` RobotProperty config `Value` still looks like this:
```
{
    kind: CircleCollision
    config: {}
}
```
The highlighted codeblock is only triggered if `config` is empty like above, and inserts default `CircleCollision` values into `Robot` properties. For any future changes to existing `CircleCollision` config, `serialize_and_change_robot_property` would be called [here](https://github.com/open-rmf/rmf_site/blob/xiyu/component_property_kinds/rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_collision.rs#L211-L216) instead.

mxgrey replied:
Right, I understood all of that context. But it stands out to me that

```rust
let Ok((ModelProperty(robot), circle_collision)) =
    params.model_descriptions.get_mut(description_entity)
else {
    return;
};
```

implies that `circle_collision` already has a specific value for this entity. If it didn't then the function would have returned early. From what I can tell, the current code **_makes an assumption_** that if the `circle_collision` component has a value while `robot.properties["collision"]["config"]` does not have an appropriate kind of value, then we are initializing the circle collision properties to their default values.

But what if, from some other part of the code, I did this to add a circle collision instead of using the UI:

```rust
world.entity_mut(model_description).insert(CircleCollision {
    radius: 100.0,
    offset: [10.0, 10.0],
});
```

If that were to happen when `CircleCollision` didn't previously exist, then we'd initialize the serialized robot property to a default value that doesn't match the actual value of the component. If instead we did

```rust
serialize_and_change_robot_property::<Collision, CircleCollision>(
    params.change_robot_property,
    circle_collision.clone(),
    robot,
    description_entity,
);
```

then we'd be putting in a matching value.

I think in general it's a bit questionable for us to be synchronizing the `ModelProperty<Robot>` component with the `CircleCollision` component inside of a widget system. That should probably be done in its own system that runs in the post-update schedule. But it would probably be best to implement this with the [relationships](https://github.com/bevyengine/bevy/pull/17398) feature, so we can keep this approach until we've upgraded to the latest bevy.

xiyuoh commented on rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_robot_properties.rs:
```diff
@@ -18,24 +18,40 @@
 /// This resource keeps track of all the properties that can be configured for a robot.
 #[derive(Resource)]
-pub struct RobotPropertyWidgets(pub HashMap<String, HashMap<String, ShowRobotPropertyWidgetFn>>);
+pub struct RobotPropertyData(
```
Oh I reverted it back to `RobotPropertyData` in this PR as the resource no longer stores widget-related callbacks. Instead it stores insert/remove component [functions](https://github.com/open-rmf/rmf_site/blob/xiyu/component_property_kinds/rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_robot_properties.rs#L48) for each `RobotPropertyKind`. The `RobotPropertyKind` widget is displayed [only if the relevant component is present](https://github.com/open-rmf/rmf_site/blob/xiyu/component_property_kinds/rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_mobility.rs#L138-L142).

The previous approach of storing the `ShowRobotPropertyWidgetFn` was limiting for this case as the property kinds widgets did not have access to world data such as Queries and Gizmos.

mxgrey replied:
I see, I still think `_Data` is misleading since I think people will assume that this resource stores the actual data to describe the robot properties.

I would argue that it does still contain widgets because the [Entity](https://github.com/open-rmf/rmf_site/blob/3c5ecf51731768c9208885feb1be6eae25433640/rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_robot_properties.rs#L46) inside of it provides the ID of a widget.

I would recommend naming it something like `RobotPropertyWidgetRegistry`.

I'd also recommend changing the tuples into concrete structs, maybe like

```rust
struct RobotPropertyWidgetRegistration {
    property_widget: Entity,
    kinds: HashMap<String, RobotPropertyKindRegistration>,
}

struct RobotPropertyKindRegistration {
    insert: InsertPropertyKindFn,
    remove: RemovePropertyKindFn,
}
```

This could make future refactors less disruptive.

xiyuoh commented on rmf_site_editor/src/site/save.rs:
```diff
@@ -1252,7 +1252,7 @@ fn generate_robots(
                 robot
                     .properties
-                    .retain(|k, v| k.is_empty() || v.as_object().is_none_or(|m| m.is_empty()));
+                    .retain(|k, v| !k.is_empty() && v.as_object().is_some());
```
Gotcha! I agree we shouldn't decide for users the Value type.

I've updated it to `!v.is_null()` in [00634b4](https://github.com/open-rmf/rmf_site/pull/265/commits/00634b422a8081749f4c61c0f57ab5f5f3e62f40) instead of removing the condition as there is a possibility of users enabling a `RobotProperty` (e.g. Mobility) for robots, but have not selected a specific `RobotPropertyKind` (e.g. DifferentialDrive) and we don't want this half-configured property to be saved to file. The pending `Value::Null` value is set [here](https://github.com/open-rmf/rmf_site/blob/xiyu/component_property_kinds/rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_robot_properties.rs#L432-L433).

mxgrey commented on rmf_site_format/src/robot.rs:
```diff
@@ -36,7 +36,7 @@ impl Default for Robot {
 }
 
 pub trait RobotProperty:
```
Does `RobotProperty` get used inside of `rmf_site_format`? If not, we could consider moving it into `rmf_site_editor` instead.

xiyuoh replied:
[b290ac5](https://github.com/open-rmf/rmf_site/pull/265/commits/b290ac56c055294497d1392f4713aeedde7671b3)

mxgrey commented on rmf_site_editor/src/site/save.rs:
```diff
@@ -1252,7 +1252,7 @@ fn generate_robots(
                 robot
                     .properties
-                    .retain(|k, v| k.is_empty() || v.as_object().is_none_or(|m| m.is_empty()));
+                    .retain(|k, v| !k.is_empty() && !v.is_null());
```
I'm not sure that we even need to check for `!v.is_null()`. I think it's plausible that some properties don't need any config structure at all, so the config field would just be `null`.

Did you run into a specific situation that needed this check?

xiyuoh replied:
Oh! You might have missed my comment [here](https://github.com/open-rmf/rmf_site/pull/265#discussion_r1951974547), it is used to filter out RobotProperties that have not been configured (i.e. no property kind selected).

mxgrey replied:
What do you think about cases where a property can simply exist or not exist without needing a configuration? At the moment we have a `{ "kind": _, "config": _ }` pattern for all properties, but I wouldn't assume that all properties need to work this way, especially user-defined properties.

In the event that a user has a custom property type that doesn't need any description, I expect they'd use `null` for its value. The current implementation would prevent that from being saved.

Even in the case of `Collision` or `Mobility` not having a `kind` selected, I think it's okay to save that they've been selected without a `kind`. We could have a diagnostic to point out when no kind is selected, and also put warnings in the log when exporting a fleet configuration where a relevant property isn't configured correctly.

xiyuoh commented:
Thanks for raising some very valid concerns! I've iterated on them and I believe I found a better approach for updating config values instead of making assumptions that are prone to bugs.

- Updated `RobotPropertyWidgetRegistry` as suggested ([2c77c7f](https://github.com/open-rmf/rmf_site/pull/265/commits/2c77c7fe43e3d580ef93350356e18bfea939eaa2)), but I took out `insert` and `remove` as I found that they're not being used anymore since we're inserting/removing components with systems instead. ([5375d09](https://github.com/open-rmf/rmf_site/pull/265/commits/5375d09d529058806092f54982aeb255c342eda6))
- In response to this [comment](https://github.com/open-rmf/rmf_site/pull/265#discussion_r1953812389), with [eeba081](https://github.com/open-rmf/rmf_site/pull/265/commits/eeba081a6c7242c768a7e9ab91448add82586e57), we now populate `config` with default values together with the `kind` field when selecting RobotPropertyKind. This removes the need to overwrite with `Kind::default()` separately from plugin widgets and possibly making unwanted re-initialization of RobotPropertyKind values. Also removed a bunch of methods enforced by the `RobotProperty` trait as they are not/no longer required by the plugins. Also, happy to look into updating the implementation with relationships when the migration is done!
- Minor updates to widgets/`save` to avoid making RobotPropertyKinds compulsory, and added a diagnostic to check for missing property kinds for cases where they do require one. ([46311eb](https://github.com/open-rmf/rmf_site/pull/265/commits/46311ebf54e5d1669ae5e8e7616b366876d7a304) and [e7f88c8](https://github.com/open-rmf/rmf_site/pull/265/commits/e7f88c8b710f6b77cf42c6a594ca420ff2f5ea15))

mxgrey commented on rmf_site_editor/src/widgets/inspector/inspect_model_description/inspect_robot_properties.rs:
```diff
@@ -15,27 +15,43 @@
 /// This resource keeps track of all the properties that can be configured for a robot.
 #[derive(Resource)]
-pub struct RobotPropertyWidgets(pub HashMap<String, HashMap<String, ShowRobotPropertyWidgetFn>>);
+pub struct RobotPropertyWidgetRegistry(pub HashMap<String, RobotPropertyWidgetRegistration>);
```
Small tip: When you have a [newtype](https://doc.rust-lang.org/rust-by-example/generics/new_types.html) like this in bevy, if you add `#[derive(Deref, DerefMut)]` then you don't need to do `.0` to access the inner data. I.e. instead of

```rust
robot_property_widgets.0.get(property)
```

you can just do

```rust
robot_property_widgets.get(property)
```

mxgrey reviewed (approved):
I think this implementation is looking really good.

One last tweak to consider would be using the [`RecallPlugin`](https://github.com/open-rmf/rmf_site/blob/a83a94b1dc5aa1bbc27d3762ef33a3989d56eba2/rmf_site_editor/src/site/recall_plugin.rs#L23-L28) for the property values. With the recall plugin you could save the last value that was used for each property (including the property kind) before it was toggled off. That way when a user switches the kind or toggles a property off and then on again, we automatically fill back in the last value they were using for it so they don't have to fill it back in. I find this useful when scrolling through different options or playing around with toggling things on/off.

You can see an example of the `Recall` trait being implemented for `AssetSource` [here](https://github.com/open-rmf/rmf_site/blob/a83a94b1dc5aa1bbc27d3762ef33a3989d56eba2/rmf_site_format/src/asset_source.rs#L134-L167).

All that being said, I don't think this is urgently needed, and this PR is already quite hefty, so I don't mind if we just make this a followup item.
